### ISSUE TITLE
ArgumentOutOfRangeException is thrown when long strings are inserted to TEXT column

### ISSUE THREAD

**Issue by kodfodrasz** [aspnet/Microsoft.Data.Sqlite#511]
I have noticed that upon insertion of long strings to a `TEXT` column `ArgumentOutOfRangeException` is thrown.

I have checked the [documentation](http://www.sqlite.org/datatype3.html), and I concluded that it should not happen, as TEXT and BLOB columns should be able to hold 1GB of data: 

> TEXT. The value is a text string, stored using the database encoding (UTF-8, UTF-16BE or UTF-16LE).
> BLOB. The value is a blob of data, stored exactly as it was input.
> ...
> Any column in an SQLite version 3 database, except an INTEGER PRIMARY KEY column, may be used to store a value of any storage class.
> ...
> Note that numeric arguments in parentheses that following the type name (ex: "VARCHAR(255)") are ignored by SQLite - SQLite does not impose any length restrictions (other than the large global SQLITE_MAX_LENGTH limit) on the length of strings, BLOBs or numeric values.

According to [referenced documentation](http://www.sqlite.org/limits.html#max_length) from the quote:

>The maximum number of bytes in a string or BLOB in SQLite is defined by the preprocessor macro SQLITE_MAX_LENGTH. The default value of this macro is 1 billion (1 thousand million or 1,000,000,000).

This lead me to investigate what may be the problem, and have also found a workaround.

I created a testsuite to reproduce the problem:

~~~csharp
using System;
using System.Text;
using Microsoft.Data.Sqlite;
using Xunit;
using Dapper;

namespace Kodfodrasz.Praegustator.Tests {
    public class SqliteBugReproduction {
        string ConnecionString = $"Data Source={Guid.NewGuid()}; Mode=Memory; Cache=Shared";

        [Fact]
        void ReproduceException() {
            var keepaliveConnection = new SqliteConnection(ConnecionString);
            keepaliveConnection.Open();

            using (var connection = new SqliteConnection(ConnecionString)) {
                string sql_create_table = @"
CREATE TABLE bug 
(
    id     GUID    NOT NULL PRIMARY KEY,
    data   TEXT    NOT NULL
)";
                connection.Execute(sql_create_table);
            }

            using (var connection = new SqliteConnection(ConnecionString)) {
                string sql_insert_row = @"
INSERT INTO bug (id, data) VALUES (@Id, @Data)
";
                var row = new {
                    Id = Guid.NewGuid(),
                    Data = new string('x', 4000)
                };

                var changed = connection.Execute(sql_insert_row, row);

                Assert.Equal(1, changed);

                var row_exception = new {
                    Id = Guid.NewGuid(),
                    Data = new string('x', 4001)
                };

                var ex = Assert.Throws<ArgumentOutOfRangeException>(() => connection.Execute(sql_insert_row, row_exception));

                Assert.Contains("Length cannot be less than zero.", ex.Message);
            }

            keepaliveConnection.Dispose();
        }


        [Fact]
        void Workaround() {
            var keepaliveConnection = new SqliteConnection(ConnecionString);
            keepaliveConnection.Open();

            using (var connection = new SqliteConnection(ConnecionString)) {
                string sql_create_table = @"
CREATE TABLE bug 
(
    id     GUID    NOT NULL PRIMARY KEY,
    data   TEXT    NOT NULL
)";
                connection.Execute(sql_create_table);
            }

            using (var connection = new SqliteConnection(ConnecionString)) {
                string sql_insert_row = @"
INSERT INTO bug (id, data) VALUES (@Id, @Data)
";
                var row = new {
                    Id = Guid.NewGuid(),
                    Data = Encoding.UTF8.GetBytes(new string('x', 999999))
                };

                var changed = connection.Execute(sql_insert_row, row);

                Assert.Equal(1, changed);
            }

            keepaliveConnection.Dispose();
        }
    }
}
~~~

I use the connection with Dapper.
Relevant `.csproj` snippets:

~~~xml
  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Dapper" Version="1.50.4" />
    <PackageReference Include="Microsoft.Data.SQLite" Version="2.1.0-preview2-final" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.0" />
    <PackageReference Include="xunit" Version="2.3.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
  </ItemGroup>
~~~

As it can be seen a viable workaround is to manually serialize to byte array, as it will be handled as a BLOB (I have noticed that largeish BLOBS were not affected while TEXTs were).

The stack trace is like:

~~~
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Microsoft.Data.Sqlite.SqliteParameterBinder.BindText(String value)
   at Microsoft.Data.Sqlite.SqliteValueBinder.Bind()
   at Microsoft.Data.Sqlite.SqliteParameter.Bind(sqlite3_stmt stmt)
   at Microsoft.Data.Sqlite.SqliteParameterCollection.Bind(sqlite3_stmt stmt)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteNonQuery()
   at Dapper.SqlMapper.ExecuteCommand(IDbConnection cnn, CommandDefinition& command, Action`2 paramReader) in C:\projects\dapper\Dapper\SqlMapper.cs:line 2827
   at Dapper.SqlMapper.ExecuteImpl(IDbConnection cnn, CommandDefinition& command) in C:\projects\dapper\Dapper\SqlMapper.cs:line 570
   at Dapper.SqlMapper.Execute(IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Nullable`1 commandTimeout, Nullable`1 commandType) in C:\projects\dapper\Dapper\SqlMapper.cs:line 443
   at Tests.SqliteBugReproduction.<>c__DisplayClass1_1.<ReproduceException>b__0() in SqliteBugReproduction.cs:line 44
   at Xunit.Assert.RecordException(Func`1 testCode) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\Record.cs:line 50
~~~

The bug also occured with normal file backed DBs.

**Comment by bricelam** [comment]
Something is setting `SqliteParamter.Size` to `-1`...

**Comment by bricelam** [comment]
We should either:
* Throw ArgumentOutOfRangeException
* Reset `Size` to its default value

**Comment by kodfodrasz** [comment]
Well, the problem could be in Dapper as well/instead. I thought it I report this to you first, and hopefully you can quickly decide the root cause.
If it is in Dapper I will report the problem there instead.

I don't wish to dig deep into this right now, tbh, as I have worked it around, and have no capacity atm. I just want to have it known by the right people.

**Comment by bricelam** [comment]
@kodfodrasz Thanks again for reporting this. It will be fixed in version 2.1.0. A size of -1 will be treated the same as on other providers: don't truncate.

**Comment by kodfodrasz** [comment]
Thank you for the quick response.

Keep u the great work,
Cheers!
